lib/protocol/http1/connection.rb

```diff
@@ -439,7 +440,14 @@ def read_response_body(method, status, headers)
 					return nil
 				end
 				
-				if (status >= 100 and status < 200) or status == 204 or status == 304
+				if status >= 100 and status < 200
```

Comment:
Maybe we should handle 101 here explicitly?

Reply:
One of my goals here is to ensure that all 1xx responses are default-safe whether or not 1xx is looped on (via `async-http`'s `Request.read`) and whether or not 101 ends up being used for a websocket connection.

The most important part is ensuring `@persistent=false` happens for all 1xx (including 101). That could be done here without the `Remainder`. Failing to do at least this much opens up a potential desync.

As I tried to mention in the OP, I'm hoping to move toward every `Response` having a body. I've been testing this in my own code with great success as it solves issues like `Statistics` and other wrappers not being able to wrap responses that lack bodies. In my own case, that includes wrapping 101's. Converting non-1xx to all have bodies is obviously a larger discussion and beyond the context here.

Returning a `Remainder` here has some benefits:

a) For 101 with `Upgrade: websocket`, if `Async::HTTP::Client` or another non-websocket-aware client is in use, then `Remainder` hints that there is something remaining on the stream (which indeed there is). More, any call to `response.close` would properly close that stream and not leave it dangling open (which is the case when `response.body==nil`).

b) For 101 with `Upgrade: <other-than-websocket>`, the `Remainder` is directly usable. Depending on the target protocol, that is potentially useful. And, the stream can still be hijacked for more direct access to the underlying socket (and the Remainder simply ignored).

c) For all 1xx, the `Remainder` can be thought of as a type of promise on the future data on the socket. Again, this hints to a developer that there's more to come. But it's more than a hint--it also clarifies that this socket is not available to be reused yet because the stream of data is not complete. To me, `nil` implies that there is nothing more coming.

Helpfully, using a `Remainder` for all 1xx shouldn't be detrimental to existing use since 1xx support is newly added. The only existing usage was via `async-websocket`, and anything that hijacks the connection was already ignoring the `nil` body, so it will simply ignore the `Remainder` body now. Barring something pretty weird, I believe this to be a non-breaking change.

All that said, if you're really uncomfortable with changing to `Remainder`, I can rework it to maintain `nil` and simply set `@persistent` instead.